validator/build.py

```diff
@@ -194,30 +194,58 @@ def GenerateValidateBin(out_dir):
   f.write("""
       var fs = require('fs');
       var path = require('path');
+      var http = require('http');
+      var https = require('https');
+      var url = require('url');
 
       function main() {
         if (process.argv.length < 3) {
-          console.error('usage: validate <file.html>');
+          console.error('usage: validate <file.html or url>');
           process.exit(1)
         }
         var args = process.argv.slice(2);
         var full_path = args[0];
-        var filename = path.basename(full_path);
-        var contents = fs.readFileSync(full_path, 'utf8');
-        var results = amp.validator.validateString(contents);
-        var output = amp.validator.renderValidationResult(results, filename);
-
-        if (output[0] === 'PASS') {
-          for (var i = 0; i < output.length; ++i) {
-            console.info(output[i]);
+        var validateFile = function(contents, filename) {
+          var results = amp.validator.validateString(contents);
+          var output = amp.validator.renderValidationResult(results, filename);
+
+          if (output[0] === 'PASS') {
+            for (var i = 0; i < output.length; ++i) {
+              console.info(output[i]);
+            }
+            process.exit(0);
+          } else {  // FAIL
+            for (var i = 0; i < output.length; ++i) {
+              console.error(output[i]);
+            }
+            process.exit(1);
           }
-          process.exit(0);
-        } else {  // FAIL
-          for (var i = 0; i < output.length; ++i) {
-            console.error(output[i]);
+        };
+
+        if (full_path.indexOf('http://') === 0 || full_path.indexOf('https://') === 0) {
+            var callback = function(response) {
+              var contents = '';
+
+              response.on('data', function (chunk) {
+                contents += chunk;
```

Comment:
I honestly don't know how much a difference it makes in practice, but think that the usual approach would be to accumulate the chunks in some array and join at the end to make sure we don't get quadratic concatenation performance on large files (and shitty Javascript interpreters).
var chunks = [];
...
    chunks.push(chunk);
...
validateFile(chunks.join(''), ...